# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/fetch/FetchService.cpp
# Commit: cf972fe0d961
# Full Hash: cf972fe0d96164788f998f04008930474526e400
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2022-02-26 21:36:36
# Regressor Bug: 1754365
# File Overlap Count: 1
# Description:
#   Bug 1754365 - Create IPC for transferring NavigationPreload's PerformanceTimingData. r=dom-worker-reviewers,jesup
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138251
# ==============================================================================

diff -r f8afd25bf41e -r cf972fe0d961 dom/fetch/FetchService.cpp
--- a/dom/fetch/FetchService.cpp	Sat Feb 26 08:40:25 2022 +0000
+++ b/dom/fetch/FetchService.cpp	Sat Feb 26 08:40:26 2022 +0000
@@ -27,6 +27,12 @@
 
 mozilla::LazyLogModule gFetchLog("Fetch");
 
+FetchServiceResponse CreateErrorResponse(nsresult aRv) {
+  IPCPerformanceTimingData ipcTimingData;
+  return MakeTuple(InternalResponse::NetworkError(aRv), ipcTimingData,
+                   EmptyString(), EmptyString());
+}
+
 // FetchInstance
 
 FetchService::FetchInstance::FetchInstance(SafeRefPtr<InternalRequest> aRequest)
@@ -129,8 +135,7 @@
   if (NS_WARN_IF(NS_FAILED(rv))) {
     FETCH_LOG(
         ("FetchInstance::Fetch FetchDriver::Fetch failed(0x%X)", (uint32_t)rv));
-    return FetchServiceResponsePromise::CreateAndResolve(
-        InternalResponse::NetworkError(rv), __func__);
+    return FetchService::NetworkErrorResponse(rv);
   }
 
   return mResponsePromiseHolder.Ensure(__func__);
@@ -147,7 +152,7 @@
   }
 
   mResponsePromiseHolder.ResolveIfExists(
-      InternalResponse::NetworkError(NS_ERROR_DOM_ABORT_ERR), __func__);
+      CreateErrorResponse(NS_ERROR_DOM_ABORT_ERR), __func__);
 }
 
 void FetchService::FetchInstance::OnResponseEnd(
@@ -156,7 +161,7 @@
   if (aReason == eAborted) {
     FETCH_LOG(("FetchInstance::OnResponseEnd end with eAborted"));
     mResponsePromiseHolder.ResolveIfExists(
-        InternalResponse::NetworkError(NS_ERROR_DOM_ABORT_ERR), __func__);
+        CreateErrorResponse(NS_ERROR_DOM_ABORT_ERR), __func__);
     return;
   }
   if (!mResponsePromiseHolder.IsEmpty()) {
@@ -174,13 +179,18 @@
   }
 
   // Get PerformanceTimingData from FetchDriver.
-  nsAutoString initiatorType;
-  nsAutoString entryName;
+  nsString initiatorType;
+  nsString entryName;
   UniquePtr<PerformanceTimingData> performanceTiming(
       mFetchDriver->GetPerformanceTimingData(initiatorType, entryName));
+  MOZ_ASSERT(performanceTiming);
+
+  FetchServiceResponse response =
+      MakeTuple(std::move(mResponse), performanceTiming->ToIPC(), initiatorType,
+                entryName);
 
   // Resolve the FetchServiceResponsePromise
-  mResponsePromiseHolder.ResolveIfExists(std::move(mResponse), __func__);
+  mResponsePromiseHolder.ResolveIfExists(std::move(response), __func__);
 }
 
 void FetchService::FetchInstance::OnResponseAvailableInternal(
@@ -216,8 +226,8 @@
 /*static*/
 RefPtr<FetchServiceResponsePromise> FetchService::NetworkErrorResponse(
     nsresult aRv) {
-  return FetchServiceResponsePromise::CreateAndResolve(
-      InternalResponse::NetworkError(aRv), __func__);
+  return FetchServiceResponsePromise::CreateAndResolve(CreateErrorResponse(aRv),
+                                                       __func__);
 }
 
 FetchService::FetchService() {